src/index.js

```diff
@@ -1,390 +1,586 @@
-            k++
-          }
-          i++
-        } else {
-          var keyedNode = oldKeyed[newKey] || []
+  return sub
+    ? oldSub ? update(sub, oldSub, dispatch) : start(sub, dispatch)
+    : oldSub ? cancel(oldSub) : oldSub
+}
+/* OR
+var toArray = function(any) {
+  return isArray(any) ? any : any == null ? [] : [any]
+}
 
-          if (oldKey === newKey) {
-            patch(element, keyedNode[0], keyedNode[1], children[k], isSvg)
-            i++
-          } else if (keyedNode[0]) {
-            patch(
-              element,
-              element.insertBefore(keyedNode[0], oldElements[i]),
-              keyedNode[1],
-              children[k],
-              isSvg
-            )
-          } else {
-            patch(element, oldElements[i], null, children[k], isSvg)
-          }
+var batch = function(subs, oldSubs, dispatch) {
+  var out = []
+  for (var i = 0; i < subs.length || i < oldSubs.length; i++) {
+    out.push(listen(subs[i], oldSubs[i], dispatch))
+  }
+  return out
+}
 
-          newKeyed[newKey] = children[k]
-          k++
-        }
+export var listen = function(sub, oldSub, dispatch) {
+  return isArray(sub) || isArray(oldSub)
+    ? batch(toArray(sub), toArray(oldSub), dispatch)
+    : sub
+      ? oldSub ? update(sub, oldSub, dispatch) : start(sub, dispatch)
+      : oldSub ? cancel(oldSub) : oldSub
+}
+*/
+
+export function app(props) {
+  var state
+  var updateInProgress
+  var view = props.view
+  var subscribe = props.subscribe
+  var subscriptions = []
+  var container = props.container
+  var element = container && container.children[0]
+  var node = element && recycleElement(element)
+
+  var setState = function(newState) {
+    if (state !== newState) {
+      state = newState
+
+      if (!updateInProgress) {
+        updateInProgress = true
+        defer(render)
       }
+    }
+  }
 
-      while (i < oldChildren.length) {
-        if (getKey(oldChildren[i]) == null) {
-          removeElement(element, oldElements[i], oldChildren[i])
-        }
-        i++
-      }
+  var dispatch = function(obj, data) {
+    typeof obj === "function"
+      ? dispatch(obj(state, data))
+      : Array.isArray(obj)
+        ? obj[1].effect(obj[1].props, dispatch, setState(obj[0]))
+        : obj != null && typeof obj.effect === "function"
+          ? dispatch([state, obj])
+          : setState(obj)
+  }
 
-      for (var i in oldKeyed) {
-        if (!newKeyed[i]) {
-          removeElement(element, oldKeyed[i][0], oldKeyed[i][1])
-        }
-      }
+  var eventProxy = function(event) {
+    dispatch(event.currentTarget.events[event.type], event)
+  }
+
+  var render = function() {
+    updateInProgress = false
+
+    if (subscribe) {
+      subscriptions = listen(subscribe(state), subscriptions, dispatch)
+    }
+
+    if (view) {
+      node = patch(container, element, node, view(state), eventProxy)
     }
-    return element
   }
+
+  dispatch(props.init)
```

Comment:
```dispatch(props.init || {})```

Providing `init` with an empty state permits to omit it in case of concise example or stateless app. We can write like:
```
app({
  view: () => (
    <div>
      Lorem ipsum dolor si amet
    </div>
  ),
  container: document.body
})
```
instead of:
```
app({
  init: () => ({ }),
  view: () => (
    <div>
      Lorem ipsum dolor si amet
    </div>
  ),
  container: document.body
})
```

Reply:
`props.init` is a function, so it should be 
```jsx
dispatch(props.init || function() {})
```
or in a more concise way
```jsx
dispatch(props.init || Function())
```

Reply:
Nevermind, `dispatch` works with objects too.

_(part of the unreadable ternary :roll_eyes: )_

Reply:
Yup.